tools/monorepo-utils/src/ci-jobs/lib/test-environment.ts

```diff
@@ -0,0 +1,201 @@
+
+/**
+ * Internal dependencies
+ */
+import { TestEnvConfigVars } from './config';
+
+/**
+ * The response for the WordPress.org stability check API.
+ */
+interface StableCheckResponse {
+	[ version: string ]: 'latest' | 'outdated' | 'insecure';
+}
+
+/**
+ * Gets all of the available WordPress versions and their associated stability.
+ *
+ * @return {Promise.<Object>} The response from the WordPress.org API.
+ */
+function getWordPressVersions(): Promise< StableCheckResponse > {
+	return new Promise< StableCheckResponse >( ( resolve, reject ) => {
+		// We're going to use the WordPress.org API to get information about available versions of WordPress.
+		const request = https.get(
+			'http://api.wordpress.org/core/stable-check/1.0/',
+			( response ) => {
+				// Listen for the response data.
+				let responseData = '';
+				response.on( 'data', ( chunk ) => {
+					responseData += chunk;
+				} );
+
+				// Once we have the entire response we can process it.
+				response.on( 'end', () =>
+					resolve( JSON.parse( responseData ) )
+				);
+			}
+		);
+
+		request.on( 'error', ( error ) => {
+			reject( error );
+		} );
+	} );
+}
+
+/**
+ * Uses the WordPress API to get the downlod URL to the latest version of an X.X version line. This
+ * also accepts "latest-X" to get an offset from the latest version of WordPress.
+ *
+ * @param {string} wpVersion The version of WordPress to look for.
+ * @return {Promise.<string>} The precise WP version download URL.
+ */
+async function getPreciseWPVersionURL( wpVersion: string ): Promise< string > {
+	const allVersions = await getWordPressVersions();
+
+	// If we're requesting a "latest" offset then we need to figure out what version line we're offsetting from.
+	const latestSubMatch = wpVersion.match( /^latest(?:-([0-9]+))?$/i );
+	if ( latestSubMatch ) {
+		for ( const version in allVersions ) {
+			if ( allVersions[ version ] !== 'latest' ) {
+				continue;
+			}
+
+			// We don't care about the patch version because we will
+			// the latest version from the version line below.
+			const versionParts = version.match( /^([0-9]+)\.([0-9]+)/ );
+
+			// We're going to subtract the offset to figure out the right version.
+			let offset = latestSubMatch[ 1 ]
+				? parseInt( latestSubMatch[ 1 ], 10 )
+				: 0;
+			let majorVersion = parseInt( versionParts[ 1 ], 10 );
+			let minorVersion = parseInt( versionParts[ 2 ], 10 );
+			while ( offset > 0 ) {
+				minorVersion--;
+				if ( minorVersion < 0 ) {
+					majorVersion--;
+					minorVersion = 9;
+				}
+				offset--;
+			}
+
+			// Set the version that we found in the offset.
+			wpVersion = majorVersion + '.' + minorVersion;
+		}
+	}
+
+	// Scan through all of the versions to find the latest version in the version line.
+	let latestVersion = null;
+	let latestPatch = -1;
+	for ( const v in allVersions ) {
+		// Parse the version so we can make sure we're looking for the latest.
+		const matches = v.match( /([0-9]+)\.([0-9]+)(?:\.([0-9]+))?/ );
```

Comment:
We should consider cutting this off at WP 6.0. I can't come up with a use case where we want to test against WP 5.9 or earlier.

Reply:
As above, I'd prefer to avoid adding more checks without a clear benefit to limiting this. If a project wants to use `latest-20` it doesn't really hurt anything (even if it's clearly ridiculous)

For what it's worth, I'd agree with you completely if it had taken extra code to get us the extended handling. This is just supported as a natural consequence of the way the function works.